File: apps/Lodgr.Api/Models/OccupancyChange.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Lodgr.Api.Models;

[Table("occupancy_changes")]
public class OccupancyChange
{
    [Key]
    [Column("change_id")]
    public long ChangeId { get; set; }

    [Column("contract_id")]
    public long ContractId { get; set; }

    [Column("effective_date")]
    public DateOnly EffectiveDate { get; set; }

    [Range(1, int.MaxValue)]
    [Column("occupants_count")]
    public int OccupantsCount { get; set; }

    [Column("reason")]
    public Reason Reason { get; set; }

    [Column("note")]
    public string? Note { get; set; }

    [Column("created_at")]
    public DateTime CreatedAt { get; set; }

    [Column("created_by")]
    public long? CreatedBy { get; set; }
}

public enum Reason
{
    MOVE_IN,
    MOVE_OUT,
    UPDATE,
    CORRECTION,
}

File: apps/api/src/main/java/io/github/hoanghonghuy/lodgr/entity/OccupancyChange.java
package io.github.hoanghonghuy.lodgr.entity;

import io.github.hoanghonghuy.lodgr.enums.OccupancyChangeReason;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "occupancy_changes")
@Getter
@Setter
public class OccupancyChange {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "change_id")
    private Long changeId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "contract_id", nullable = false)
    private Contract contract;

    @Column(name = "effective_date", nullable = false)
    private LocalDate effectiveDate;

    @Column(name = "occupants_count", nullable = false)
    private Integer occupantsCount;

    @Enumerated(EnumType.STRING)
    @Column(name = "reason", nullable = false)
    private OccupancyChangeReason reason;

    @Column(name = "note")
    private String note;

    @Column(name = "created_at", nullable = false)
    private LocalDateTime createdAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by")
    private User createdBy;
}
